lib/src/winrt/foundation/collections/ikeyvaluepair.dart

```diff
@@ -0,0 +1,278 @@
+class IKeyValuePair<K, V> extends IInspectable {
+  // vtable begins at 6, is 2 entries long.
+  final V Function(Pointer<COMObject>)? _creator;
+  final V Function(int)? _enumCreator;
+
+  /// Creates an instance of [IKeyValuePair] using the given [ptr].
+  ///
+  /// [K] must be of type `GUID`, `int`, `Object`, `String`, or `WinRTEnum`
+  /// (e.g. `PedometerStepKind`).
+  ///
+  /// [V] must be of type `Object`, `String`, or `WinRT` (e.g. `IJsonValue`,
+  /// `ProductLicense`).
+  ///
+  /// [creator] must be specified if [V] is a `WinRT` type.
+  /// ```dart
+  /// final keyValuePair =
+  ///     IKeyValuePair<String, IJsonValue?>.fromRawPointer(ptr,
+  ///         creator: IJsonValue.fromRawPointer);
+  /// ```
+  ///
+  /// [enumCreator] must be specified if [V] is a `WinRTEnum` type.
+  /// ```dart
+  /// final keyValuePair =
+  ///     IKeyValuePair<String, ChatMessageStatus?>.fromRawPointer(ptr,
+  ///         enumCreator: ChatMessageStatus.from);
+  /// ```
+  IKeyValuePair.fromRawPointer(
+    super.ptr, {
+    V Function(Pointer<COMObject>)? creator,
+    V Function(int)? enumCreator,
+  })  : _creator = creator,
+        _enumCreator = enumCreator {
+    if (!isSupportedKeyValuePair<K, V>()) {
+      throw ArgumentError('Unsupported key-value pair: IKeyValuePair<$K, $V>');
+    }
+
+    if (isSubtypeOfInspectable<V>() && creator == null) {
+      throw ArgumentError.notNull('creator');
+    }
+
+    if (isSubtypeOfWinRTEnum<V>() && enumCreator == null) {
+      throw ArgumentError.notNull('enumCreator');
+    }
+  }
+
+  /// Gets the key of the key-value pair.
+  K get key {
+    if (isSameType<K, GUID>()) return _key_GUID as K;
+    if (isSameType<K, int>()) return _key_Uint32 as K;
+    if (isSameType<K, PedometerStepKind>()) {
+      return _key_PedometerStepKind as K;
+    }
+    if (isSameType<K, String>()) return _key_String as K;
+
+    return _key_Object;
+  }
+
+  GUID get _key_GUID {
+    final retValuePtr = calloc<GUID>();
+
+    final hr = ptr.ref.lpVtbl.value
+        .elementAt(6)
+        .cast<
+            Pointer<NativeFunction<HRESULT Function(Pointer, Pointer<GUID>)>>>()
+        .value
+        .asFunction<
+            int Function(
+                Pointer, Pointer<GUID>)>()(ptr.ref.lpVtbl, retValuePtr);
+
+    if (FAILED(hr)) throw WindowsException(hr);
+
+    return retValuePtr.ref;
+  }
+
+  int get _key_Uint32 {
+    final retValuePtr = calloc<Uint32>();
+
+    try {
+      final hr = ptr.ref.lpVtbl.value
+          .elementAt(6)
+          .cast<
+              Pointer<
+                  NativeFunction<HRESULT Function(Pointer, Pointer<Uint32>)>>>()
+          .value
+          .asFunction<
+              int Function(
+                  Pointer, Pointer<Uint32>)>()(ptr.ref.lpVtbl, retValuePtr);
+
+      if (FAILED(hr)) throw WindowsException(hr);
+
+      return retValuePtr.value;
+    } finally {
+      free(retValuePtr);
+    }
+  }
+
+  PedometerStepKind get _key_PedometerStepKind {
```

Comment:
Here are the supported key-value pairs of `IKeyValuePair` according to IIDs taken from [here](https://github.com/Elmue/WindowsRT-GUID-Analyzer/blob/main/Release/Windows%2011%20(Build%2022000)%20WinRT%20Interfaces.ini):
https://github.com/timsneath/win32/blob/339c6e4ac04a9d2f2c1933c958bc0893be65dee2/lib/src/winrt_constants.dart#L62-L148

According to these, there is only one enum type (`PedometerStepKind`) used as a key type of `IKeyValuePair`. It seemed easier to handle it this way instead of adding a new parameter (`keyCreator`) that takes an enum constructor for the key type of `ÌKeyValuePair` which will allow me to create an instance of that enum type. 

I can modify the implementation to handle these types with the `keyCreator` parameter I mentioned above if that's what you want.

Reply:
I wonder if the best thing is to make the `PedometerStepKind` pieces an extension method in winrt/device/sensors? I don't _know_ that this is the right answer, but it does feel like a very odd outlier. Are there other classes that might have a similar pattern, I wonder?

Reply:
> I wonder if the best thing is to make the `PedometerStepKind` pieces an extension method in winrt/device/sensors?

I've created extension methods for the `PedometerStepKind` pieces but I'm not sure if it's the way you wanted it. Could you take a look?

> Are there other classes that might have a similar pattern, I wonder?

I don't think there are.